keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md

```diff
@@ -0,0 +1,289 @@
+storageVersionHash in the discovery document. These fact make the storage
+migrator not working properly ([details][]).
+
+[storageVersionHash]:https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L979
+[details]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md#ha-masters
+
+We propose a mechanism for HA API servers to vote on the storage version to use,
+then show the selected storage version in the discovery document and switch to
+use the selected version atomically.
+
+## API changes
+
+### Storage Version API
+
+We introduce a new API `StorageVersion`, in a new API group
+`kube-apiserver.internal.k8s.io/v1alpha1`. 
+
+```golang
+// Storage version for a specific resource.
+type StorageVersion struct {
+  TypeMeta
+  // The name is the GVR, i.e., <group>.<version>.<resource>.
+  ObjectMeta
+  // The selected storage version. All API server intances encode the resource
+  // objects in this version when committing it to etcd.
+  Selectedversion string
+  // Proposed storage version candidate, keyed by the ID of the proposing API
+  // server.
+  CandidateVersions []VersionCandidate
+}
+
+type VersionCandidate struct {
+  // The ID of the reporting API server. 
+  APIServerID string
+  // The preferred storage version of the reporting API server. The preferred
+  // storage version is hardcoded in each Kubernetes release.
+  PreferredVersion string
+}
+```
+
+We will discuss how the `StorageVersion` is updated and used in the [Changes to API
+servers](#changes-to-api-servers) section.
+
+## Changes to API servers
+
+### Curating a list of participating API servers in HA master
+
+Currently, API servers maintain such a list in the "kubernetes" endpoints. See
+this [design doc][] for details.
+
+We will formalize the API in another KEP. For the purpose of this KEP, all we
+need to know is that HA master is able to maintain a list of participating API
+servers in etcd.
+
+[design doc]:https://github.com/kubernetes/community/pull/939
+
+### Voting for Storage Version
+
+During bootstrap, for each resource, the kube-apiserver 
+* gets the StorageVersion object for this resource,
+  * if the object does not exist, creates one, setting this kube-apiserver's
+  preferred storage version as the Selectedversion, also adding its preferred
+  storage version to the CandidateVersions list. Then jumps to the last step.
+* gets the list of participating API servers,
+* updates the StorageVersion locally. Specifically,
+  * creates or updates the CandidateVersions. 
+  * checks if there is any version candidate whose APIServerID is not a
+    participating API server. This means the API server has gone and the entry
+    is stale. Removes such entries.
+  * checks if all participating API server votes for the same version. If so,
+    sets the version as the Selectedversion.
```

Comment:
What if not all of them are voting for the same version?

Reply:
The preferred version in the first vote becomes the first selected version.

Afterwards, only after all apiservers prefer a new version, then it becomes the new selected version.